File: online-election/online-election.cpp
class TopVotedCandidate {
private:
    map<int, int> leadAtTime {};
    
public:
    TopVotedCandidate(vector<int>& persons, vector<int>& times) {
        
        int lead = INT_MIN;
        
        unordered_map<int, int> CandidateVotes;
        
        for (int i = 0; i < times.size(); ++i) {
            
            ++CandidateVotes[persons[i]];
                                     
            lead = (CandidateVotes[persons[i]] >= CandidateVotes[lead]) ? persons[i] : lead;
            
            leadAtTime[times[i]] = lead;
        }
    }
    
    int q(int t) {
        
        // use map::upper_bound to find the first time AFTER t
        // then go back by one index and return that value
        // this ensures that the returned value comes from the last recorded time <= t
        
        return (--leadAtTime.upper_bound(t))->second;
    }
};

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate* obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj->q(t);
 */

File: online-election/online-election.java
class TopVotedCandidate {
    
    private HashMap<Integer, Integer> leadAtTime = new HashMap();
    
    private int[] times;

    public TopVotedCandidate(int[] persons, int[] times) {
        
        this.times = times;
        
        int lead = Integer.MIN_VALUE;
        
        HashMap<Integer, Integer> candidateVotes = new HashMap();
        
        for (int i = 0; i < times.length; ++i) {
            
            candidateVotes.put(persons[i], candidateVotes.getOrDefault(persons[i], 0) + 1);
            
            if (i == 0 || candidateVotes.get(persons[i]) >= candidateVotes.get(lead)) { lead = persons[i]; }
            
            leadAtTime.put(times[i], lead);
        }
    }
    
    public int q(int t) {
        
        int left = 0, right = times.length - 1, mid;
        
        while (left < right) {
            
            mid = left + ((right - left) / 2) + 1;
            
            if (times[mid] <= t) {
                
                left = mid;
            }
            else {
                
                right = mid - 1;
            }
        }
        
        return leadAtTime.get(times[left]);
    }
}

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj.q(t);
 */